near-sdk/src/store/unordered_map/mod.rs

```diff
@@ -0,0 +1,545 @@
+mod impls;
+mod iter;
+
+use std::borrow::Borrow;
+use std::{fmt, mem};
+
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::crypto_hash::{CryptoHasher, Sha256};
+use crate::{env, IntoStorageKey};
+
+pub use entry::{Entry, OccupiedEntry, VacantEntry};
+
+pub use self::iter::{Iter, IterMut, Keys, Values, ValuesMut};
+use super::bucket::BucketIndex;
+use super::{Bucket, LookupMap, ERR_INCONSISTENT_STATE};
+
+const ERR_NOT_EXIST: &str = "Key does not exist in map";
+
+/// A non-iterable, lazily loaded storage map that stores its content directly on the storage trie.
+///
+/// This map stores the values under a hash of the map's `prefix` and [`BorshSerialize`] of the key
+/// using the map's [`CryptoHasher`] implementation.
+///
+/// The default hash function for [`UnorderedMap`] is [`Sha256`] which uses a syscall
+/// (or host function) built into the NEAR runtime to hash the key. To use a custom function,
+/// use [`with_hasher`]. Alternative builtin hash functions can be found at
+/// [`near_sdk::crypto_hash`](crate::crypto_hash).
+///
+/// # Examples
+/// ```
+/// use near_sdk::store::UnorderedMap;
+///
+/// // Initializes a map, the generic types can be inferred to `UnorderedMap<String, u8, Sha256>`
+/// // The `b"a"` parameter is a prefix for the storage keys of this data structure.
+/// let mut map = UnorderedMap::new(b"a");
+///
+/// map.insert("test".to_string(), 7u8);
+/// assert!(map.contains_key("test"));
+/// assert_eq!(map.get("test"), Some(&7u8));
+///
+/// let prev = std::mem::replace(&mut map["test"], 5u8);
+/// assert_eq!(prev, 7u8);
+/// assert_eq!(map["test"], 5u8);
+/// ```
+///
+/// `UnorderedMap` also implements an [`Entry API`](Self::entry), which allows
+/// for more complex methods of getting, setting, updating and removing keys and
+/// their values:
+///
+/// ```
+/// use near_sdk::store::UnorderedMap;
+///
+/// // type inference lets us omit an explicit type signature (which
+/// // would be `UnorderedMap<String, u8>` in this example).
+/// let mut player_stats = UnorderedMap::new(b"m");
+///
+/// fn random_stat_buff() -> u8 {
+///     // could actually return some random value here - let's just return
+///     // some fixed value for now
+///     42
+/// }
+///
+/// // insert a key only if it doesn't already exist
+/// player_stats.entry("health".to_string()).or_insert(100);
+///
+/// // insert a key using a function that provides a new value only if it
+/// // doesn't already exist
+/// player_stats.entry("defence".to_string()).or_insert_with(random_stat_buff);
+///
+/// // update a key, guarding against the key possibly not being set
+/// let stat = player_stats.entry("attack".to_string()).or_insert(100);
+/// *stat += random_stat_buff();
+/// ```
+///
+/// [`with_hasher`]: Self::with_hasher
+pub struct UnorderedMap<K, V, H = Sha256>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    keys: Bucket<K>,
+    values: LookupMap<K, ValueAndIndex<V>, H>,
```

Comment:
:thinking: hm, I'd naively expect that we us the following repr here: 

```
struct UnordereMap {
  lm: LookupMap<CryptoHash, (K, V)>
}
```

ie, that we just use hashes as keys and key-value pairs as values. 

Which also makes me think, that maybe lookup map is the wrong abstaction? Ie, the difference between lookup map and unordered map is that the former uses borsh serialize as the trie key, while the latter uses sha hash as the key.

So perhaps we need to introduce some kind of RawSet internal abstraction that just stores byte keys and values and parametrized by how we get thouse bytes out of Rust types. 

Than lookup map would be one specialization that uses Borsh for both, and UnorderedMap would use Borsh + CryptoHasher.

Reply:
Yeah, I guess I should make a more careful pass over it, as it seems I got myself confused about what this is doing at all :)

Reply:
Opened https://github.com/near/near-sdk-rs/pull/596 which might make things more clear and be a higher leverage piece of code to weigh in on